File: @sfa/sdk/safety.ts
import { ExitCode } from "./types";
import { exitWithError, emitProgress } from "./output";

/**
 * Safety state for the current agent invocation.
 * Read from environment on startup, used for enforcement.
 */
export interface SafetyState {
  /** Current invocation depth */
  depth: number;
  /** Maximum allowed depth */
  maxDepth: number;
  /** Call chain (agent names) up to and including this agent */
  callChain: string[];
  /** Session ID for this invocation tree */
  sessionId: string;
}

/**
 * Initialize safety state from environment and CLI flags.
 * Reads SFA_DEPTH, SFA_MAX_DEPTH, SFA_CALL_CHAIN, SFA_SESSION_ID.
 * Generates a new session ID if this is a top-level invocation.
 */
export function initSafety(
  agentName: string,
  maxDepthFlag?: number,
): SafetyState {
  const depth = parseInt(process.env.SFA_DEPTH ?? "0", 10);
  const maxDepth =
    maxDepthFlag ??
    parseInt(process.env.SFA_MAX_DEPTH ?? "5", 10);
  const parentChain = process.env.SFA_CALL_CHAIN
    ? process.env.SFA_CALL_CHAIN.split(",")
    : [];
  const sessionId = process.env.SFA_SESSION_ID ?? crypto.randomUUID();

  // Check for loop: if this agent already appears in the call chain, refuse
  if (parentChain.includes(agentName)) {
    const loopPath = [...parentChain, agentName].join(" → ");
    exitWithError(
      `Loop detected: ${loopPath}\nAgent "${agentName}" already appears in the call chain.`,
      ExitCode.FAILURE,
    );
  }

  // Append this agent to the call chain
  const callChain = [...parentChain, agentName];

  // Propagate to process.env so child processes and subsystems can read them
  process.env.SFA_DEPTH = String(depth);
  process.env.SFA_MAX_DEPTH = String(maxDepth);
  process.env.SFA_CALL_CHAIN = callChain.join(",");
  process.env.SFA_SESSION_ID = sessionId;

  return { depth, maxDepth, callChain, sessionId };
}

/**
 * Check whether spawning a subagent is allowed at the current depth.
 * Throws if at or above max depth.
 */
export function checkDepthLimit(safety: SafetyState): void {
  if (safety.depth + 1 >= safety.maxDepth) {
    throw new Error(
      `Maximum invocation depth reached (${safety.maxDepth}). ` +
      `Cannot spawn subagent at depth ${safety.depth + 1}.`,
    );
  }
}

/**
 * Check whether spawning a named subagent would create a loop.
 * Throws if the agent name already appears in the call chain.
 */
export function checkLoop(safety: SafetyState, targetAgent: string): void {
  if (safety.callChain.includes(targetAgent)) {
    const loopPath = [...safety.callChain, targetAgent].join(" → ");
    throw new Error(`Loop detected: ${loopPath}`);
  }
}

/**
 * Build SFA_* environment variables for a subagent invocation.
 * Increments depth, appends to call chain, forwards session ID.
 */
export function buildSubagentSafetyEnv(safety: SafetyState): Record<string, string> {
  return {
    SFA_DEPTH: String(safety.depth + 1),
    SFA_MAX_DEPTH: String(safety.maxDepth),
    SFA_CALL_CHAIN: safety.callChain.join(","),
    SFA_SESSION_ID: safety.sessionId,
  };
}

/**
 * Set up timeout with an AbortController.
 * Returns the controller and a cleanup function.
 */
export function setupTimeout(
  agentName: string,
  timeoutSeconds: number,
): { controller: AbortController; cleanup: () => void } {
  const controller = new AbortController();

  const timer = setTimeout(() => {
    emitProgress(agentName, `timeout after ${timeoutSeconds}s`);
    controller.abort();
  }, timeoutSeconds * 1000);

  return {
    controller,
    cleanup: () => clearTimeout(timer),
  };
}

/**
 * Set up signal handlers (SIGINT, SIGTERM).
 * Returns a cleanup function to remove them.
 */
export function setupSignalHandlers(
  agentName: string,
  controller: AbortController,
): { cleanup: () => void } {
  const onSigint = () => {
    emitProgress(agentName, "interrupted");
    controller.abort();
    setTimeout(() => process.exit(ExitCode.SIGINT), 100);
  };

  const onSigterm = () => {
    emitProgress(agentName, "terminating");
    controller.abort();
    setTimeout(() => process.exit(ExitCode.SIGTERM), 5000);
  };

  process.on("SIGINT", onSigint);
  process.on("SIGTERM", onSigterm);

  return {
    cleanup: () => {
      process.off("SIGINT", onSigint);
      process.off("SIGTERM", onSigterm);
    },
  };
}

File: sdk/golang/sfa/safety.go
package sfa

import (
	"context"
	"crypto/rand"
	"fmt"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"
)

// SafetyState tracks invocation depth, call chain, and session identity.
type SafetyState struct {
	Depth     int
	MaxDepth  int
	CallChain []string
	SessionID string
}

// initSafety reads SFA_* safety env vars, performs loop detection, and propagates state.
func initSafety(agentName string, maxDepthFlag int) (*SafetyState, error) {
	depth := parseInt(os.Getenv("SFA_DEPTH"), 0)
	maxDepth := parseInt(os.Getenv("SFA_MAX_DEPTH"), maxDepthFlag)

	// Parse call chain
	chainStr := os.Getenv("SFA_CALL_CHAIN")
	var chain []string
	if chainStr != "" {
		chain = strings.Split(chainStr, ",")
	}

	// Loop detection — check if this agent is already in the call chain
	for _, name := range chain {
		if name == agentName {
			chain = append(chain, agentName)
			return nil, fmt.Errorf("loop detected: %s", strings.Join(chain, " → "))
		}
	}

	// Append current agent to call chain
	chain = append(chain, agentName)

	// Session ID — generate if top-level
	sessionID := os.Getenv("SFA_SESSION_ID")
	if sessionID == "" {
		sessionID = generateUUID()
	}

	safety := &SafetyState{
		Depth:     depth,
		MaxDepth:  maxDepth,
		CallChain: chain,
		SessionID: sessionID,
	}

	// Propagate to process environment
	os.Setenv("SFA_DEPTH", fmt.Sprintf("%d", depth))
	os.Setenv("SFA_MAX_DEPTH", fmt.Sprintf("%d", maxDepth))
	os.Setenv("SFA_CALL_CHAIN", strings.Join(chain, ","))
	os.Setenv("SFA_SESSION_ID", sessionID)

	return safety, nil
}

// checkDepthLimit returns an error if depth+1 would exceed maxDepth.
func checkDepthLimit(safety *SafetyState) error {
	if safety.Depth+1 >= safety.MaxDepth {
		return fmt.Errorf("depth limit reached: current depth %d, max depth %d", safety.Depth, safety.MaxDepth)
	}
	return nil
}

// checkLoop returns an error if the target agent is already in the call chain.
func checkLoop(safety *SafetyState, targetAgent string) error {
	for _, name := range safety.CallChain {
		if name == targetAgent {
			chain := append(safety.CallChain, targetAgent)
			return fmt.Errorf("loop detected: %s", strings.Join(chain, " → "))
		}
	}
	return nil
}

// buildSubagentSafetyEnv returns env vars with incremented depth for subagent invocation.
func buildSubagentSafetyEnv(safety *SafetyState) map[string]string {
	return map[string]string{
		"SFA_DEPTH":      fmt.Sprintf("%d", safety.Depth+1),
		"SFA_MAX_DEPTH":  fmt.Sprintf("%d", safety.MaxDepth),
		"SFA_CALL_CHAIN": strings.Join(safety.CallChain, ","),
		"SFA_SESSION_ID": safety.SessionID,
	}
}

// setupTimeout returns a context with a timeout and a cancel function.
func setupTimeout(agentName string, timeoutSeconds int) (context.Context, context.CancelFunc) {
	if timeoutSeconds <= 0 {
		return context.WithCancel(context.Background())
	}
	return context.WithTimeout(context.Background(), time.Duration(timeoutSeconds)*time.Second)
}

// setupSignalHandlers installs SIGINT and SIGTERM handlers that cancel the context.
// Returns a cleanup function that removes the signal handlers.
func setupSignalHandlers(agentName string, cancel context.CancelFunc) func() {
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		sig := <-sigCh
		cancel()

		switch sig {
		case syscall.SIGINT:
			emitProgress(agentName, "interrupted (SIGINT)")
			// Give a moment for cleanup, then exit
			time.Sleep(100 * time.Millisecond)
			os.Exit(ExitSIGINT)
		case syscall.SIGTERM:
			emitProgress(agentName, "terminated (SIGTERM)")
			time.Sleep(5 * time.Second) // grace period
			os.Exit(ExitSIGTERM)
		}
	}()

	return func() {
		signal.Stop(sigCh)
		close(sigCh)
	}
}

// generateUUID produces a UUID v4 string.
func generateUUID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 2
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
